File: Data Structures & Algorithms/largest-rectangle-in-histogram/submission-10.js
class Solution {
    /**
     * @param {number[]} heights
     * @return {number}
     */
    largestRectangleArea(heights) { 
        let stack = [] 
        let max = 0
        for (let i = 0; i < heights.length ; i ++) { 
            let start = i
            while(stack.length > 0 && stack[stack.length - 1][1] > heights[i]) {
                let [ index, height ] = stack.pop() 
                max = Math.max(max, height * (i - index)) 
                start = index
            } 

            stack.push([start, heights[i]])
        } 

        for (const [ index, height ] of stack) {
            max = Math.max(max, height * (heights.length - index))
        } 

        return max
    }
}

File: Data Structures & Algorithms/validate-parentheses/submission-10.py
class Solution:
    def isValid(self, s: str) -> bool:
        close_bracket = { 
            ']': '[', 
            '}': "{", 
            ')': '('
        } 

        stack = [] 

        for c in s: 
            if c in close_bracket:  
                if len(stack) == 0: 
                    return False 
                    
                open_bracket = stack.pop() 
                if close_bracket[c] != open_bracket: 
                    return False 
            else: 
                stack.append(c) 

        return len(stack) == 0